File: day9/soln.py
def solve1(ipt, print_result=True, **kwargs):
    ipt = [int(i) for i in ipt]
    step = 5 if kwargs.get('is_test') else 25

    invalid = 0
    for i in range(step, len(ipt)):
        target = ipt[i]
        seen = set()
        valid = False

        for j in range(i-step, i):
            candidate = ipt[j]

            if target - candidate in seen and target != target - candidate:
                valid = True
                break
            else:
                seen.add(candidate)

        if not valid:
            invalid = target

            if print_result:
                print(target)

            return invalid

def solve2(ipt, **kwargs):
    invalid = solve1(ipt, False, **kwargs)
    ipt = [int(i) for i in ipt]

    for i in range(len(ipt)):
        total = ipt[i]
        ran = [ipt[i]]
        for j in range(i+1, len(ipt)):
            ran.append(ipt[j])
            total += ipt[j]

            if total == invalid:
                print((min(ran) + max(ran)))
                return
            elif total > invalid:
                break

File: day9/soln.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func solve(ipt []int, step int) {
	var invalid int

	for i := step; i < len(ipt); i++ {
		target := ipt[i]
		seen := make(map[int]bool)
		valid := false

		for j := i-step; j < i; j++ {
			candidate := ipt[j]

			if seen[target-candidate] && target != target - candidate {
				valid = true
				break
			} else {
				seen[candidate] = true
			}
		}

		if !valid {
			invalid = target
			fmt.Printf("part 1: %d\n", target)
		}
	}

	for i := 0; i < len(ipt); i++ {
		var ran []int
		total := ipt[i]
		ran = append(ran, ipt[i])

		for j := i+1; j < len(ipt); j++ {
			ran = append(ran, ipt[j])
			total += ipt[j]

			if total == invalid {
				min := 0
				max := 0
				for _, r := range ran {
					if min == 0 || r < min {
						min = r
					} else if r > max {
						max = r
					}
				}

				fmt.Printf("part 2: %d\n", min + max)
				return
			} else if total > invalid {
				break
			}
		}
	}
}

func main() {
	var ipt []int
	test, _ := os.Open("test_input.txt")
	defer test.Close()

	scanner := bufio.NewScanner(test)
	for scanner.Scan() {
		asInt, _ := strconv.Atoi(scanner.Text())
		ipt = append(ipt, asInt)
	}

	fmt.Println("TEST")
	solve(ipt, 5)
	fmt.Println("")

	ipt = nil
	real, _ := os.Open("input.txt")
	defer real.Close()

	scanner = bufio.NewScanner(real)
	for scanner.Scan() {
		asInt, _ := strconv.Atoi(scanner.Text())
		ipt = append(ipt, asInt)
	}

	fmt.Println("RESULT")
	solve(ipt, 25)
	fmt.Println("")
}